### ISSUE TITLE
cacti 1.1.9  and clog_webapi.php permission issue

### ISSUE THREAD

**Issue by mortenstevens** [Cacti/cacti#791]
Hi,

After upgrading to 1.1.9 I see multiple warnings/errors like this:

Do we need to give cacti 1.1.9 read/write permissions on /lib/clog_webapi.php?

2017/06/05 15:17:54 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 334 CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:17:54 - ERROR PHP WARNING: Invalid argument supplied for foreach() in file: /usr/share/cacti/lib/clog_webapi.php on line: 334
2017/06/05 15:17:54 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 332 scandir)(CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:17:54 - ERROR PHP WARNING: scandir(): (errno 13): Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 332
2017/06/05 15:17:54 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 332 scandir)(CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:17:54 - ERROR PHP WARNING: scandir(/usr/share/cacti/log): failed to open dir: Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 332
2017/06/05 15:15:26 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 334 CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:15:26 - ERROR PHP WARNING: Invalid argument supplied for foreach() in file: /usr/share/cacti/lib/clog_webapi.php on line: 334
2017/06/05 15:15:26 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 332 scandir)(CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:15:26 - ERROR PHP WARNING: scandir(): (errno 13): Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 332
2017/06/05 15:15:26 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 332 scandir)(CactiErrorHandler)(/lib/functions.php: 4359 cacti_debug_backtrace)
2017/06/05 15:15:26 - ERROR PHP WARNING: scandir(/usr/share/cacti/log): failed to open dir: Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 332

**Comment by reboot1983** [comment]
Is your cacti.log written to /usr/share/cacti/log ?
This directory needs ownership or read/write access

**Comment by mortenstevens** [comment]
@reboot1983 

Thanks for your quick answer. We are using /var/log/cacti.log for Fedora and we have a symlink from /usr/share/cacti/log to /var/log/cacti

I guess your patch should fix this issue?

**Comment by mortenstevens** [comment]
Looks like a selinux issue...

type=AVC msg=audit(1496687064.110:1627): avc:  denied  { read } for  pid=5344 comm="httpd" name="cacti" dev="dm-0" ino=4326705 scontext=system_u:system_r:httpd_t:s0 tcontext=system_u:object_r:httpd_log_t:s0 tclass=dir permissive=0

But I did not understand why this happens only with 1.1.8 and 1.1.9? With all cacti releases before we had no problem to keep the log files in /var/log/cacti/cacti.log...

Any ideas?

**Comment by reboot1983** [comment]
Since this release we have a feature for reading rotated logfiles. So we scan the dir for logfiles.

**Comment by reboot1983** [comment]
Not sure if the patch solves it for you. Noticed a possible bug if the cacti logpath is not set in the settings.

**Comment by mortenstevens** [comment]
@reboot1983 

The patch doesn't help. I guess the new feature to scan for rotated logfiles causes the selinux permission issue... If if turn selinux off (setenforce 0) all works fine... But we need to get it ready with selinux enabled for Fedora...

**Comment by reboot1983** [comment]
Do you have set ownership of the cacti.log file only? Or the complete directory?

**Comment by mortenstevens** [comment]
@reboot1983 

Yes. The directory permission should be okay since it's working fine with turned off selinux (setenforce 0)

directory permissio /var/log/
drwxrwxr-x. cacti  apache system_u:object_r:httpd_log_t:s0 cacti

file permissions in the /var/log/cacti directory:
-rw-rw-r--. cacti apache system_u:object_r:httpd_log_t:s0 cacti.log
-rw-rw-r--. cacti apache system_u:object_r:httpd_log_t:s0 cacti.log-20170329.gz
-rw-rw-r--. cacti apache system_u:object_r:httpd_log_t:s0 cacti.log-20170401.gz
-rw-rw-r--. cacti apache system_u:object_r:httpd_log_t:s0 cacti.log-20170501.gz
-rw-rw-r--. cacti apache system_u:object_r:httpd_log_t:s0 cacti.log-20170601.gz

If guess it's only about the new feature to scan the dir for logfiles. Maybe we can get a switch to turn this feature off?

**Comment by reboot1983** [comment]
I'll ask @cigamit if I can/may add an option in the settings panel (settings, tab paths?) for this.

**Comment by mortenstevens** [comment]
@reboot1983 
Maybe it would be better in the configuration file or turn it off by default since it seems to be conflict with selinux.

**Comment by cigamit** [comment]
Resolved.

**Comment by reboot1983** [comment]
@mortenstevens I was looking into this issue, but this really seems a permission issue for the log directory.

I noticed a bug in the listing of the files. You showed me your directory contents, and there are rotated files present. They are .gz files and because these files are compressed, we need to exclude those in the listing. But also this one won't solve your problem.

You mentioned that, when disabling selinux, it's working, so I guess there needs to be more/different access to the log directory. Are you able to investigate this and gently play with some permissions?

https://unix.stackexchange.com/questions/50639/httpd-cant-write-to-folder-file-because-of-selinux

Second:
I've changed the clog code in this way, that if we cannot read the logdirectory for it's files, I'm only showing cacti.log and suppress the scandir error. I believe this one should fix your problem. I've attached a patch for this. Are you able to test this?

**Comment by mortenstevens** [comment]
@reboot1983 

I think it's more a selinux issue. After changing the selinux extended file attribute for /var/log/cacti to httpd_sys_rw_content_t it's working fine. 

As an interim solution until selinux-policy is updated I changed the rpm spec file to httpd_sys_rw_content_t for /var/log/cacti

https://src.fedoraproject.org/cgit/rpms/cacti.git/commit/?id=040b494892f6a5f5110243399eee4cae17c4c427

**Comment by reboot1983** [comment]
Cool! Thanks for your feedback.

**Comment by mortenstevens** [comment]
@reboot1983 

Just a small question: httpd_sys_rw_content_t is not the best solution for us, because we need httpd_log_t otherwise logrotation fails because selinux blocks access to files with httpd_sys_rw_content_t.

So I have another question: Is it just possible to remove these PHP warnings in the cacti log file for the case for cacti installations without access to scan the log directory? I think it's not intended to generate PHP warnings for just an optional feature which scans for rotated logfiles? What do you think? I would be very happy if it is possible to remove these php error warnings with the next cacti release.

**Comment by reboot1983** [comment]
I've solved this last week in 1.1.11. When Cacti doesn't have read rights, it will no longer throw an error, but instead it will only show cacti.log in the CLOG.

**Comment by mortenstevens** [comment]
@reboot1983 
I tried commit a310dc63fdd94480422507286e7c690c42e26fc2 on a 1.1.0 installation but I still get these two PHP errors if I navigate to the cacti log page:

2017/06/30 14:04:27 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 337 scandir)(CactiErrorHandler)(/lib/functions.php: 4363 cacti_debug_backtrace)
2017/06/30 14:04:27 - ERROR PHP WARNING: scandir(): (errno 13): Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 337
2017/06/30 14:04:27 - CMDPHP PHP ERROR WARNING Backtrace: (/clog.php: 39 clog_view_logfile)(/lib/clog_webapi.php: 172 filter)(/lib/clog_webapi.php: 337 scandir)(CactiErrorHandler)(/lib/functions.php: 4363 cacti_debug_backtrace)
2017/06/30 14:04:27 - ERROR PHP WARNING: scandir(/usr/share/cacti/log): failed to open dir: Permission denied in file: /usr/share/cacti/lib/clog_webapi.php on line: 337

**Comment by cigamit** [comment]
I just pushed a change.  Please see if this issue still occurs and let us know.

**Comment by mortenstevens** [comment]
@cigamit 
Confirmed. The latest commit 2a66c3aa44d69e6a82a5a2b26fd6f99aeb849de6 seems to fix the PHP error warnings in the cacti log file :) Thank you.

I got only the selinux avc denied message:

type=AVC msg=audit(1498825971.204:484): avc:  denied  { read } for  pid=2308 comm="httpd" name="cacti" dev="dm-0" ino=4326705 scontext=system_u:system_r:httpd_t:s0 tcontext=system_u:object_r:httpd_log_t:s0 tclass=dir permissive=0

But maybe this needs to be fixed by in the selinux-policy. Or do we have the possibility to avoid scanning the cacti logdir completely?
